src/Symfony/Component/Messenger/EventListener/ResetMemoryUsageListener.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Messenger\Event\WorkerMessageFailedEvent;
+use Symfony\Component\Messenger\Event\WorkerMessageHandledEvent;
+use Symfony\Component\Messenger\Event\WorkerMessageReceivedEvent;
+
+/**
+ * @author Tim Düsterhus <tim@tideways-gmbh.com>
+ */
+class ResetMemoryUsageListener implements EventSubscriberInterface
+{
+    public function resetBefore(WorkerMessageReceivedEvent $event): void
+    {
+        // Reset the peak memory usage for accurate measurement of the
+        // memory usage on a per-message basis.
+        memory_reset_peak_usage();
+    }
+
+    public function collectAfter(WorkerMessageHandledEvent|WorkerMessageFailedEvent $event): void
+    {
+        // Run the cycle collector after handling a message to avoid it
+        // running while the message is reserved and in processing.
+        gc_collect_cycles();
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            WorkerMessageReceivedEvent::class => ['resetBefore', -1024],
+            WorkerMessageFailedEvent::class => ['collectAfter', -1024],
+            WorkerMessageHandledEvent::class => ['collectAfter', -1024],
```

Comment:
I suggest doing the collection only when the worker is idle. That's a good enough default to me. Ppl that need more frequent GC can still register their listener if they really want to:
```suggestion
            WorkerRunningEvent::class => ['collectAfter', -1024],
```